// @vue-macros/short-bind@0.2.3/dist/chunk-IG3N6B67.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunk7ADJIZ7Djs = require('./chunk-7ADJIZ7D.js');

// src/index.ts
var _common = require('@vue-macros/common');
var name = "unplugin-vue-short-bind";
function rollup(options = {}) {
  let api;
  return {
    name,
    enforce: "pre",
    configResolved(config) {
      try {
        api = _common.getVuePluginApi.call(void 0, config.plugins);
      } catch (e) {
      }
    },
    buildStart(rollupOpts) {
      if (api === void 0)
        try {
          api = _common.getVuePluginApi.call(void 0, rollupOpts.plugins);
        } catch (error) {
          this.warn(error);
          return;
        }
      if (!api)
        return;
      api.options.template ||= {};
      api.options.template.compilerOptions ||= {};
      api.options.template.compilerOptions.nodeTransforms ||= [];
      api.options.template.compilerOptions.nodeTransforms.push(
        _chunk7ADJIZ7Djs.transformShortBind.call(void 0, options)
      );
    }
  };
}
var src_default = {
  rollup,
  vite: rollup
};



exports.src_default = src_default;
