// @vue-macros/define-prop@0.3.4/dist/chunk-SDGVU76K.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }



var _chunkCOYLF2XRjs = require('./chunk-COYLF2XR.js');

// src/index.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
var _unplugin = require('unplugin');






var _common = require('@vue-macros/common');
var _api = require('@vue-macros/api');
function resolveOptions(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  const include = _common.getFilterPattern.call(void 0, 
    [_common.FilterFileType.VUE_SFC_WITH_SETUP, _common.FilterFileType.SETUP_SFC],
    framework
  );
  return {
    include,
    isProduction: _process2.default.env.NODE_ENV === "production",
    edition: "kevinEdition",
    ...options,
    version
  };
}
var name = "unplugin-vue-define-prop";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    const { resolve, handleHotUpdate } = _api.RollupResolve.call(void 0, );
    return {
      name,
      enforce: "pre",
      buildStart() {
        if (framework === "rollup" || framework === "vite") {
          _api.setResolveTSFileIdImpl.call(void 0, resolve(this));
        }
      },
      resolveId(id) {
        if (id === _common.normalizePath.call(void 0, _chunkCOYLF2XRjs.helperId))
          return id;
      },
      loadInclude(id) {
        return _common.normalizePath.call(void 0, id) === _chunkCOYLF2XRjs.helperId;
      },
      load(id) {
        if (_common.normalizePath.call(void 0, id) === _chunkCOYLF2XRjs.helperId)
          return _chunkCOYLF2XRjs.code_default;
      },
      transformInclude: filter,
      transform(code, id) {
        return _chunkCOYLF2XRjs.transformDefineProp.call(void 0, 
          code,
          id,
          options.edition,
          options.isProduction
        );
      },
      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction;
        },
        handleHotUpdate
      }
    };
  }
);



exports.src_default = src_default;
