// @vue-macros/volar@0.18.17/dist/export-props.js
// src/export-props.ts
var import_language_core2 = require("@vue/language-core");
var import_pluginutils = require("@rollup/pluginutils");

// src/common.ts
var import_language_core = require("@vue/language-core");
function getVueLibraryName(vueVersion) {
  return vueVersion < 2.7 ? "@vue/runtime-dom" : "vue";
}
function addProps(content, decl, vueLibName) {
  (0, import_language_core.replaceAll)(
    content,
    /setup\(\) {/g,
    "props: ({} as ",
    ...decl,
    "),\n",
    "setup() {"
  );
  content.push(
    `type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;
`,
    `type __VLS_TypePropsToRuntimeProps<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? { type: import('${vueLibName}').PropType<__VLS_NonUndefinedable<T[K]>> } : { type: import('${vueLibName}').PropType<T[K]>, required: true } };
`
  );
  return true;
}
function getVolarOptions(vueCompilerOptions) {
  return vueCompilerOptions.vueMacros;
}

// src/export-props.ts
function transform({
  file,
  sfc,
  ts,
  vueLibName,
  volarOptions,
  fileName
}) {
  const filter = (0, import_pluginutils.createFilter)(
    volarOptions.include || /.*/,
    volarOptions.exclude
  );
  if (!filter(fileName))
    return;
  const props = /* @__PURE__ */ Object.create(null);
  let changed = false;
  for (const stmt of sfc.scriptSetup.ast.statements) {
    if (!ts.isVariableStatement(stmt))
      continue;
    const exportModifier = stmt.modifiers?.find(
      (m) => m.kind === ts.SyntaxKind.ExportKeyword
    );
    if (!exportModifier)
      continue;
    const start = exportModifier.getStart(sfc.scriptSetup?.ast);
    const end = exportModifier.getEnd();
    (0, import_language_core2.replaceSourceRange)(file.content, "scriptSetup", start, end);
    changed = true;
    for (const decl of stmt.declarationList.declarations) {
      if (!ts.isIdentifier(decl.name))
        continue;
      props[decl.name.text] = !!decl.initializer;
    }
  }
  if (changed) {
    addProps(
      file.content,
      [
        `__VLS_TypePropsToRuntimeProps<{
${Object.entries(props).map(([prop, optional]) => `  ${prop}${optional ? "?" : ""}: typeof ${prop}`).join(",\n")}
  }>`
      ],
      vueLibName
    );
  }
}
var plugin = ({
  modules: { typescript: ts },
  vueCompilerOptions
}) => {
  return {
    name: "vue-macros-export-props",
    version: 1,
    resolveEmbeddedFile(fileName, sfc, embeddedFile) {
      if (embeddedFile.kind !== import_language_core2.FileKind.TypeScriptHostFile || !sfc.scriptSetup || !sfc.scriptSetup.ast)
        return;
      const vueLibName = getVueLibraryName(vueCompilerOptions.target);
      transform({
        file: embeddedFile,
        sfc,
        vueLibName,
        ts,
        fileName,
        volarOptions: getVolarOptions(vueCompilerOptions)?.exportProps || {}
      });
    }
  };
};
module.exports = plugin;
