// @vue-macros/volar@0.18.17/dist/setup-jsdoc.js
// src/setup-jsdoc.ts
var import_language_core = require("@vue/language-core");
function hasJSDocNodes(node) {
  if (!node)
    return false;
  const { jsDoc } = node;
  return !!jsDoc && jsDoc.length > 0;
}
function transform({
  file,
  sfc,
  ts
}) {
  let jsDoc;
  if (hasJSDocNodes(sfc.scriptSetup.ast.statements[0])) {
    jsDoc = sfc.scriptSetup.ast.statements[0].jsDoc.at(-1);
  }
  for (const stmt of sfc.scriptSetup.ast.statements) {
    if (!ts.isExportAssignment(stmt))
      continue;
    if (hasJSDocNodes(stmt))
      jsDoc ??= stmt.jsDoc.at(-1);
  }
  if (jsDoc) {
    (0, import_language_core.replace)(
      file.content,
      /(?=export\sdefault)/,
      `${sfc.scriptSetup?.content.slice(jsDoc.pos, jsDoc.end)}
`
    );
  }
}
var plugin = ({ modules: { typescript: ts } }) => {
  return {
    name: "vue-macros-setup-jsdoc",
    version: 1,
    resolveEmbeddedFile(fileName, sfc, embeddedFile) {
      if (embeddedFile.kind !== import_language_core.FileKind.TypeScriptHostFile || !sfc.scriptSetup || !sfc.scriptSetup.ast)
        return;
      transform({
        file: embeddedFile,
        sfc,
        ts
      });
    }
  };
};
module.exports = plugin;
