// @vue-macros/better-define@1.7.4/dist/chunk-FH46YAD3.mjs
import {
  transformBetterDefine
} from "./chunk-CZPXXJOC.mjs";

// src/index.ts
import process from "node:process";
import { createUnplugin } from "unplugin";
import {
  FilterFileType,
  createFilter,
  detectVueVersion,
  getFilterPattern
} from "@vue-macros/common";
import { RollupResolve, setResolveTSFileIdImpl } from "@vue-macros/api";
function resolveOptions(options, framework) {
  const version = options.version || detectVueVersion();
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework
  );
  return {
    include,
    isProduction: process.env.NODE_ENV === "production",
    ...options,
    version
  };
}
var name = "unplugin-vue-better-define";
var src_default = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = createFilter(options);
    const { resolve, handleHotUpdate } = RollupResolve();
    return {
      name,
      enforce: "pre",
      buildStart() {
        if (framework === "rollup" || framework === "vite") {
          setResolveTSFileIdImpl(resolve(this));
        }
      },
      transformInclude: filter,
      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction);
        } catch (error) {
          this.warn(`${name} ${error}`);
          console.warn(error);
        }
      },
      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction;
        },
        handleHotUpdate
      }
    };
  }
);

export {
  src_default
};
