// @vue-macros/setup-block@0.3.3/dist/chunk-ISP6R33L.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkVBD4TI52js = require('./chunk-VBD4TI52.js');

// src/index.ts
var _unplugin = require('unplugin');





var _common = require('@vue-macros/common');
function resolveOptions(options) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC],
    defaultLang: "ts",
    ...options,
    version
  };
}
var name = "unplugin-vue-setup-block";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}) => {
    const options = resolveOptions(userOptions);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        return _chunkVBD4TI52js.transformSetupBlock.call(void 0, code, id, options.defaultLang);
      }
    };
  }
);



exports.src_default = src_default;
