// @vue-macros/define-render@1.5.3/dist/chunk-4INPJ3ZE.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkTSJ2UQINjs = require('./chunk-TSJ2UQIN.js');

// src/index.ts
var _unplugin = require('unplugin');





var _common = require('@vue-macros/common');
function resolveOptions(options) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [
      _common.REGEX_VUE_SFC,
      _common.REGEX_SETUP_SFC,
      /\.(vue|setup\.[cm]?[jt]sx?)\?vue/
    ],
    version,
    ...options
  };
}
var name = "unplugin-vue-define-render";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}) => {
    const options = resolveOptions(userOptions);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "post",
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        return _chunkTSJ2UQINjs.transformDefineRender.call(void 0, code, id);
      }
    };
  }
);



exports.src_default = src_default;
