// eslint-plugin-import-x@0.4.4/lib/rules/no-self-import.js
"use strict";
const eslint_compat_utils_1 = require("eslint-compat-utils");
const utils_1 = require("../utils");
function isImportingSelf(context, node, requireName) {
    const filePath = (0, eslint_compat_utils_1.getPhysicalFilename)(context);
    if (filePath !== '<text>' && filePath === (0, utils_1.resolve)(requireName, context)) {
        context.report({
            node,
            messageId: 'self',
        });
    }
}
module.exports = (0, utils_1.createRule)({
    name: 'no-self-import',
    meta: {
        type: 'problem',
        docs: {
            category: 'Static analysis',
            description: 'Forbid a module from importing itself.',
            recommended: true,
        },
        schema: [],
        messages: {
            self: 'Module imports itself.',
        },
    },
    defaultOptions: [],
    create(context) {
        return (0, utils_1.moduleVisitor)((source, node) => {
            isImportingSelf(context, node, source.value);
        }, { commonjs: true });
    },
});
