// @vue-macros/jsx-directive@0.8.12/dist/chunk-ZER4OZWK.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkOLLFTUSEjs = require('./chunk-OLLFTUSE.js');

// src/index.ts
var _unplugin = require('unplugin');






var _common = require('@vue-macros/common');
function resolveOptions(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  const include = _common.getFilterPattern.call(void 0, 
    [_common.FilterFileType.VUE_SFC_WITH_SETUP, _common.FilterFileType.SRC_FILE],
    framework
  );
  return {
    include,
    exclude: [_common.REGEX_NODE_MODULES],
    ...options,
    version
  };
}
var name = "unplugin-vue-jsx-directive";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      transformInclude: filter,
      transform(code, id) {
        return _chunkOLLFTUSEjs.transformJsxDirective.call(void 0, code, id, options.version);
      }
    };
  }
);



exports.src_default = src_default;
