// @vue-macros/jsx-directive@0.8.12/dist/chunk-3G5TRYRD.mjs
import {
  transformJsxDirective
} from "./chunk-4KNWYT6X.mjs";

// src/index.ts
import { createUnplugin } from "unplugin";
import {
  FilterFileType,
  REGEX_NODE_MODULES,
  createFilter,
  detectVueVersion,
  getFilterPattern
} from "@vue-macros/common";
function resolveOptions(options, framework) {
  const version = options.version || detectVueVersion();
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SRC_FILE],
    framework
  );
  return {
    include,
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version
  };
}
var name = "unplugin-vue-jsx-directive";
var src_default = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = createFilter(options);
    return {
      name,
      enforce: "pre",
      transformInclude: filter,
      transform(code, id) {
        return transformJsxDirective(code, id, options.version);
      }
    };
  }
);

export {
  src_default
};
