// @vue-macros/define-models@1.2.3/dist/chunk-UAWOJ5DZ.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});






var _chunk7FRPXCOYjs = require('./chunk-7FRPXCOY.js');

// src/index.ts
var _unplugin = require('unplugin');






var _common = require('@vue-macros/common');
function resolveOptions(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  const include = _common.getFilterPattern.call(void 0, 
    [_common.FilterFileType.VUE_SFC_WITH_SETUP, _common.FilterFileType.SETUP_SFC],
    framework
  );
  return {
    include,
    unified: true,
    ...options,
    version
  };
}
var name = "unplugin-vue-define-models";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (_common.normalizePath.call(void 0, id).startsWith(_chunk7FRPXCOYjs.helperPrefix))
          return id;
      },
      loadInclude(id) {
        return _common.normalizePath.call(void 0, id).startsWith(_chunk7FRPXCOYjs.helperPrefix);
      },
      load(_id) {
        const id = _common.normalizePath.call(void 0, _id);
        if (id === _chunk7FRPXCOYjs.emitHelperId)
          return _chunk7FRPXCOYjs.emit_helper_default;
        else if (id === _chunk7FRPXCOYjs.useVmodelHelperId)
          return _chunk7FRPXCOYjs.use_vmodel_default;
      },
      transformInclude: filter,
      transform(code, id) {
        return _chunk7FRPXCOYjs.transformDefineModels.call(void 0, code, id, options.version, options.unified);
      }
    };
  }
);



exports.src_default = src_default;
