// @vue-macros/setup-sfc@0.17.3/dist/chunk-4RJFPEPL.mjs
import {
  hotUpdateSetupSFC,
  transformSetupSFC
} from "./chunk-AVJOJVUV.mjs";

// src/index.ts
import { createUnplugin } from "unplugin";
import {
  REGEX_SETUP_SFC_SUB,
  createFilter,
  detectVueVersion
} from "@vue-macros/common";
function resolveOptions(options) {
  const version = options.version || detectVueVersion();
  return {
    include: [REGEX_SETUP_SFC_SUB],
    exclude: [/vitest\.setup\.\w+$/],
    ...options,
    version
  };
}
var name = "unplugin-vue-setup-sfc";
var src_default = createUnplugin(
  (userOptions = {}) => {
    const options = resolveOptions(userOptions);
    const filter = createFilter(options);
    return {
      name,
      enforce: "pre",
      transformInclude: filter,
      transform: transformSetupSFC,
      vite: {
        config() {
          return {
            esbuild: {
              exclude: options.include,
              include: options.exclude
            }
          };
        },
        handleHotUpdate: (ctx) => {
          if (filter(ctx.file)) {
            return hotUpdateSetupSFC(ctx, filter);
          }
        }
      }
    };
  }
);

export {
  src_default
};
