// @vue-macros/setup-sfc@0.17.3/dist/chunk-Y5WYLBAY.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _chunk5MVWTFYZjs = require('./chunk-5MVWTFYZ.js');

// src/index.ts
var _unplugin = require('unplugin');




var _common = require('@vue-macros/common');
function resolveOptions(options) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_SETUP_SFC_SUB],
    exclude: [/vitest\.setup\.\w+$/],
    ...options,
    version
  };
}
var name = "unplugin-vue-setup-sfc";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}) => {
    const options = resolveOptions(userOptions);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      transformInclude: filter,
      transform: _chunk5MVWTFYZjs.transformSetupSFC,
      vite: {
        config() {
          return {
            esbuild: {
              exclude: options.include,
              include: options.exclude
            }
          };
        },
        handleHotUpdate: (ctx) => {
          if (filter(ctx.file)) {
            return _chunk5MVWTFYZjs.hotUpdateSetupSFC.call(void 0, ctx, filter);
          }
        }
      }
    };
  }
);



exports.src_default = src_default;
